spark/src/main/scala/com/cloudera/livy/spark/client/ClientSession.scala

```diff
@@ -80,6 +99,22 @@ class ClientSession(val sessionId: Int, createRequest: CreateClientRequest)
     new JobStatus(id, status.state, status.result, status.error)
   }
 
+
+  private def copyResourceToHDFS(dataStream: InputStream, name: String): URI = {
+    val filePath = new Path(sessionHome, name)
+    val outFile = fs.create(filePath, true)
+    val buffer = new Array[Byte](512 * 1024)
+    var read = -1
+    try {
+      while ({read = dataStream.read(buffer); read != -1}) {
```

Comment:
I wonder if there's a short NIO-based version to replace this, to avoid the in-memory buffer...